src/app/fdctl/run/run.c

```diff
@@ -417,83 +232,143 @@ main_pid_namespace( void * args ) {
 
-    child_pids[ child_cnt ] = clone_tile( config, tile, tile_to_cpu[ i ], floating_cpu_set );
+    int pipefd[ 2 ];
+    if( FD_UNLIKELY( pipe2( pipefd, O_CLOEXEC ) ) ) FD_LOG_ERR(( "pipe2() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+    fds[ 1+child_cnt ] = (struct pollfd){ .fd = pipefd[ 0 ], .events = 0 };
+    child_pids[ child_cnt ] = execve_tile( tile, tile_to_cpu[ i ], floating_cpu_set, config_memfd, pipefd[ 1 ] );
+    if( FD_UNLIKELY( close( pipefd[ 1 ] ) ) ) FD_LOG_ERR(( "close() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
     strncpy( child_names[ child_cnt ], fd_topo_tile_kind_str( tile->kind ), 32 );
     child_cnt++;
   }
 
+  if( FD_UNLIKELY( close( config_memfd ) ) ) FD_LOG_ERR(( "close() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( close( fd_log_private_shared_memfd ) ) ) FD_LOG_ERR(( "close() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+
   if( FD_UNLIKELY( sched_setaffinity( 0, sizeof(cpu_set_t), floating_cpu_set ) ) )
     FD_LOG_ERR(( "sched_setaffinity failed (%i-%s)", errno, fd_io_strerror( errno ) ));
 
+  if( FD_UNLIKELY( args->pipefd[ 1 ]!=1 ) ) FD_LOG_ERR(( "unexpected fd %d for main pipe", fds[ 0 ].fd ));
+  if( FD_UNLIKELY( fd_log_private_fileno!=4 ) ) FD_LOG_ERR(( "unexpected fd %d for logfile", fd_log_private_fileno ));
+
+  int allow_fds[ 3+FD_TOPO_MAX_TILES ];
+  allow_fds[ 0 ] = 1; /* write end of main pipe */
+  allow_fds[ 1 ] = 2; /* stderr */
+  allow_fds[ 2 ] = 4; /* logfile */
+  for( ulong i=0; i<child_cnt; i++ )
+    allow_fds[ 3+i ] = fds[ 1+i ].fd; /* read end of child pipes */
+
   struct sock_filter seccomp_filter[ 128UL ];
   populate_sock_filter_policy_pidns( 128UL, seccomp_filter );
 
-  int allow_fds[] = {
-    2, /* stderr */
-    3, /* logfile */
-  };
- 
   fd_sandbox( config->development.sandbox,
               config->uid,
               config->gid,
-              sizeof(allow_fds)/sizeof(allow_fds[ 0 ]),
+              1+child_cnt, /* RLIMIT_NOFILE needs to be set to the nfds argument of poll() */
+              3+child_cnt, /* Number of fds we are allowing to be mapped into the process */
               allow_fds,
               sock_filter_policy_pidns_instr_cnt,
               seccomp_filter );
 
-  /* we are now the init process of the pid namespace. if the init process
-     dies, all children are terminated. If any child dies, we terminate the
-     init process, which will cause the kernel to terminate all other children
-     bringing all of our processes down as a group. */
-  int wstatus;
-  pid_t exited_pid = wait4( -1, &wstatus, (int)__WCLONE, NULL );
-  if( FD_UNLIKELY( exited_pid == -1 ) ) {
-    fd_log_private_fprintf_nolock_0( STDERR_FILENO, "wait4() failed (%i-%s)", errno, fd_io_strerror( errno ) );
-    exit_group( 1 );
-  }
+  /* No more access to the log file, write log lines only to STDERR from
+     here out. */
 
-  char * name = "unknown";
-  ulong tile_idx = ULONG_MAX;
   for( ulong i=0; i<child_cnt; i++ ) {
```

Comment:
The point of this loop is just to print more friendly error messages about failure on boot, right? Otherwise isn't there a race between this code and the tile boot?